### PR TITLE
Revamp using new PETSc_jll binaries

### PR DESCRIPTION
**Pull request by simonbyrne**

### CODE REVIEW

File: src/const.jl [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    return datasize_ref[]
+end
+
+
+const PETSC_REAL = DataTypeFromString("Real")
+const PETSC_SCALAR = DataTypeFromString("Scalar")
+
+const PetscReal =
+    PETSC_REAL == PETSC_DOUBLE ? Cdouble :
+    PETSC_REAL == PETSC_FLOAT ? Cfloat :
```
Comment (stevengj) [inline src/const.jl:56]: Previously, we supported both single- and double-precision versions simultaneously.   You just compile multiple Petsc libraries and call them separately via `ccall` depending on the precision.

Comment (simonbyrne) [reply src/const.jl:56]: I do plan on supporting that: the first challenge is building the libraries. I've opened an issue here: https://github.com/JuliaPackaging/Yggdrasil/issues/1527

Comment (simonbyrne) [reply src/const.jl]: I've updated to support this, once the necessary libraries are built.

**Comment by simonbyrne** [comment]
@stevengj would you have any objection to me merging this? I realise it's not complete (in terms of functionality vs the old code), but it would make it easier for us to do more incremental development.

**Comment by simonbyrne** [comment]
@JaredCrean2 as well

**Comment by thomasgibson** [comment]
@simonbyrne This has been a great start! I tried to see how far we can push the specification of solver options. Here is an example using CG plus PETSc's native AMG preconditioner: https://github.com/JuliaParallel/PETSc.jl/commit/20265160bad973eefe79ddb4fccc55fef58099aa 

I think there is some issue with nesting options prefixes - or perhaps I am missing something?

**Comment by stevengj** [comment]
That's fine with me.

Are you planning to manually translate the Petsc headers rather than using Clang.jl?

**Comment by simonbyrne** [comment]
I think so: the bigger challenge is keeping track of object lifetimes and figuring out correct return objects, which generated wrappers can't help with.

**Comment by stevengj** [comment]
Yes, that seems reasonable to me.   My initial plan was to do manual header translation, but the Petsc guys got upset with that suggestion and felt that we couldn't possibly keep up with their advanced API changes (see JuliaLang/julia#2645).

**Pull request merged**